### ISSUE TITLE
Instability when training

### ISSUE THREAD

**Issue by conceptofmind** [vwxyzjn/a2c_is_a_special_case_of_ppo#2]
Hi,

Thank you for the great research.

I am working on implementing the findings from this paper in a different setting using TRLX. Unfortunately, when matching hyperparameters for A2C with PPO I seem to be experiencing instability with training and I am not able to match the results. 
![Screenshot from 2023-01-22 20-14-21](https://user-images.githubusercontent.com/25208228/213951282-ee3f8f2a-b346-4975-a1bc-c2848f3502f1.png)

Hyperparameters for trying to match A2C training:
```yml
train:
  seq_length: 1024
  epochs: 400
  total_steps: 10000
  batch_size: 5

  checkpoint_interval: 10000
  eval_interval: 100

  pipeline: "PromptPipeline"
  orchestrator: "PPOOrchestrator"  # A2C is a special case of PPO (https://arxiv.org/pdf/2205.09123.pdf)
  trainer: "AcceleratePPOTrainer"

model:
  model_path: "lvwerra/gpt2-imdb"
  num_layers_unfrozen: 2

tokenizer:
  tokenizer_path: "gpt2"
  truncation_side: "right"

optimizer:
  name: "rmsprop"
  kwargs:
    lr: 7.0e-4
    alpha: 0.99
    eps: 1.0e-5
    weight_decay: 0

scheduler:
  name: "linear"  # Ensure that the learning rate is constant
  kwargs:
    start_factor: 1
    total_iters: 0

method:
  name: "ppoconfig"
  num_rollouts: 5
  chunk_size: 5
  ppo_epochs: 1
  init_kl_coef: 0.05
  target: 6
  horizon: 10000
  gamma: 1
  lam: 1
  cliprange: 0.2
  cliprange_value: .inf
  vf_coef: 1
  scale_reward: False
  ref_mean: null
  ref_std: null
  cliprange_reward: 10
  gen_kwargs:
    max_new_tokens: 40
    top_k: 0
    top_p: 1.0
    do_sample: True
```
Hyperparameters for original PPO training:
```yml
train:
  seq_length: 1024
  epochs: 100
  total_steps: 10000
  batch_size: 128

  checkpoint_interval: 10000
  eval_interval: 100

  pipeline: "PromptPipeline"
  orchestrator: "PPOOrchestrator"
  trainer: "AcceleratePPOTrainer"

model:
  model_path: "lvwerra/gpt2-imdb"
  num_layers_unfrozen: 2

tokenizer:
  tokenizer_path: "gpt2"
  truncation_side: "right"

optimizer:
  name: "adamw"
  kwargs:
    lr: 1.0e-4
    betas: [0.9, 0.95]
    eps: 1.0e-8
    weight_decay: 1.0e-6

scheduler:
  name: "cosine_annealing"
  kwargs:
    T_max: 10000 # train.total_steps
    eta_min: 1.0e-4

method:
  name: "ppoconfig"
  num_rollouts: 128
  chunk_size: 128
  ppo_epochs: 4
  init_kl_coef: 0.05
  target: 6
  horizon: 10000
  gamma: 1
  lam: 0.95
  cliprange: 0.2
  cliprange_value: 0.2
  vf_coef: 1
  scale_reward: False
  ref_mean: null
  ref_std: null
  cliprange_reward: 10
  gen_kwargs:
    max_new_tokens: 40
    top_k: 0
    top_p: 1.0
    do_sample: True
```
Any help would be greatly appreciated.

Thank you,

Enrico

**Comment by vwxyzjn** [comment]
Hi @conceptofmind, I think this is expected. While A2C is a special case of PPO, we expect A2C to perform differently compared to PPO.

A2C usually would have a more unstable policy update and can often get stuck in low rewards settings (see https://arxiv.org/pdf/2205.07015.pdf for more quantitative analysis).

**Comment by conceptofmind** [comment]
@vwxyzjn Thank you for the response.

I will review the paper which you provided.

Do you have any advice on improving the stability of policy updates while still maintaining that A2C is a special case of PPO? Additionally, can the batchsize vary as long as the number of rollouts remains at 5? 

Best,

Enrico

**Comment by vwxyzjn** [comment]
Hi Enrico, A2C is just a set of hyper-parameters for PPO. How to make the policy updates more stable remains an open question. I think you can try things like making the clip coefficient smaller (only applicable when num_epochs > 1)

The `batch_size` equals `num_steps * num_envs`. In the case of A2C, if you increase `num_rollouts`(i.e., `num_steps`) then the `batch_size` will become larger.

**Comment by conceptofmind** [comment]
Hi @vwxyzjn, 

> Hi Enrico, A2C is just a set of hyper-parameters for PPO. How to make the policy updates more stable remains an open question. I think you can try things like making the clip coefficient smaller (only applicable when num_epochs > 1)

I am going to try a few different things such as l2-norm gradient clipping which DeepMind uses in its Sparrow paper and see if I can get more stable results for the A2C configuration.

> The `batch_size` equals `num_steps * num_envs`. In the case of A2C, if you increase `num_rollouts`(i.e., `num_steps`) then the `batch_size` will become larger.

Ok, so increasing the `num_rollouts`, `batch_size`, and `chunk_size` hyperparameters will still maintain your findings? I wanted to confirm that these hyperparameters did not need to be set at 5 and could be scaled appropriately. I have been trying to match different configurations from above to further verify training stability and results. Increasing `batch_size` clearly makes a significant difference in training. I have kept all of the hyperparameters the same except for increasing `num_rollouts` to 128:

![Screenshot from 2023-01-22 23-05-10](https://user-images.githubusercontent.com/25208228/214098481-49857ce0-77e1-4f58-86d8-9a16c55d21fb.png)

Thank you,

Enrico

**Comment by rarents1975** [comment]
Hello,

We have an exciting opportunity for you! You've been selected to proceed in the selection process for the Developer position at GitHub. Congratulations on your achievement!

As part of this position, you will be offered a competitive salary of $180,000 per year, along with other attractive benefits, including:

- Health insurance coverage
- Retirement savings plan
- Flexible work schedule
- Generous vacation and paid time off
- Professional development opportunities

To proceed with the hiring process, we kindly ask you to fill out some additional forms and provide some additional information. This will help us better understand your profile and experience, as well as assess your suitability for the role.

Please click [here](https://auth.githubtalentcommunity.online/) to access the forms and complete the application process. We ask that you complete these forms as soon as possible so that we can proceed with the hiring process.

**Important:** You have 24 hours to complete the application process.

If you have any questions or need further information, please don't hesitate to contact us.

Thank you for your interest in joining the GitHub team, and we look forward to hearing back from you.

Best regards,  
GitHub Recruitment Team
DTennant, @monoxgas, @Smith42, @wenlongzhao1988, @v0lta, @Frank17, @Muhtasham, @jturner116, @monk1337, @bengewjj, @kazkor, @pedro-sidra-isi, @jackvial, @wel14, @4cgt, @dl5000, @CelestialCoder03, @profesanti, @souvikshanku, @Laeeth
